libarchive/archive_read_support_format_rar5.c

```diff
@@ -3911,6 +3911,13 @@ static int do_unpack(struct archive_read* a, struct rar5* rar,
 			case GOOD:
 				/* fallthrough */
 			case BEST:
+				/* No data is returned here. But because a sparse-file aware
+				 * caller (like archive_read_data_into_fd) may treat zero-size
+				 * as a sparse file block, we need to update the offset
+				 * accordingly. At this point the decoder doesn't have any
+				 * pending uncompressed data blocks, so the current position in
+				 * the output file should be last_write_ptr. */
+				if (offset) *offset = rar->cstate.last_write_ptr;
```

Comment:
Thanks for the PR!

Do you think it would be better to place this update into the `rar5_read_data` function directly after the update to the `size` variable? E.g.:

```c
static int rar5_read_data(struct archive_read *a, const void **buff,
    size_t *size, int64_t *offset) {
	int ret;
	struct rar5* rar = get_context(a);

	if (size)
		*size = 0;

	if (offset)
		*offset = rar->cstate.last_write_ptr;  // ** here

	if(rar->file.dir > 0) {
		/* Don't process any data if this file entry was declared
		 * as a directory. This is needed, because entries marked as
```

This way the `offset` and `size` variables would be updated in the same place.

Reply:
I thought about it initially. And yes, it should work. But it's kind of not obvious why the default offset at the beginning of `rar 5_read_data` should (or may) be `rare->c state.last_write_ptr`.
In proper implementation the default value for `*offset` should be something like `last_returned_offset + last_returned_size`. In most cases it doesn't equal `last_write_ptr`.

So yes, it will be nice to have `size` and `offset` initialized in the same place, but moving this fix to `rar5_read_data` function 'as is' will not be an honest solution. I don't really know what would be better.